orttraining/orttraining/core/graph/allreduce_optimizer_graph_builder.cc

```diff
@@ -163,11 +162,19 @@ Status AllreduceOptimizerGraphBuilder::BuildInternal(
   // check if all gradients are finite
   ArgDef global_grad_norm_argdef;
   ArgDef global_grad_norm_finite_argdef;
-  if (opt_graph_config_.use_mixed_precision) {
+
+  if (opt_graph_config_.enable_grad_norm_clip ||
+      (opt_graph_config_.use_mixed_precision &&
+       opt_graph_config_.mixed_precision_type == MixedPrecisionDataType::FP16)) {
+    //gradient norm for bfloat16 is not ready yet. skip it to unblock the testing
```

Comment:
Why not just use `enable_grad_norm_clip` to control everything about gradient norm? We can set `enable_grad_norm_clip` in ConfigureForTraining based on precision mode.

Reply:
In the case of fp16 run, user can still set enable_grad_norm_clip to false and we should not use gradient norm in the optimizer computation. 

So we can't override enable_grad_norm_clip for this purpose.

Reply:
I should write it clearer. Inside this function, a single boolean flag is enough for controlling if gradient norm is computed. You can move
```
my_flag = opt_graph_config_.enable_grad_norm_clip ||
      (opt_graph_config_.use_mixed_precision &&
       opt_graph_config_.mixed_precision_type == MixedPrecisionDataType::FP16
```
outside this function and pass `my_flag` into this function.

---
In reply to: [520930473](https://github.com/microsoft/onnxruntime/pull/5728#discussion_r520930473) [](ancestors = 520930473)

Reply:
done